File: func.java
import java.util.*;
import java.io.*;

public class func 
{
	public static void main(String[] args) throws IOException
	{
		Scanner scan = new Scanner(System.in);
		System.out.print("Input absolute path to dataset (.csv): ");
		String path = scan.nextLine();
		
		Scanner fin = new Scanner(new File(path));
		ArrayList<Double> X = new ArrayList<Double>();
		ArrayList<Double> Y = new ArrayList<Double>();
		
		while(fin.hasNext())
		{
			String[] coords = fin.nextLine().split(",");
			X.add(Double.parseDouble(coords[0]));
			Y.add(Double.parseDouble(coords[1]));
		}
		System.out.println(X);
		System.out.println(Y);

		ArrayList<Double> theta = new ArrayList<Double>();
		ArrayList<Double> coeff = new ArrayList<Double>();
		do {
			theta.add(0.0);
			coeff = gradDescent(theta, X, Y, 0.000001, 0.001);
		} while(coeff.get(0) == Double.MAX_VALUE);

		System.out.println(coeff);
	}

	public static ArrayList<Double> gradDescent(ArrayList<Double> theta, ArrayList<Double> X, ArrayList<Double> Y, double err, double alpha)
	{
		ArrayList<Double> ret = new ArrayList<Double>();

		ArrayList<Double> cost_hist = new ArrayList<Double>();
		double J = Double.MAX_VALUE;
		cost_hist.add(0.0);
		cost_hist.add(J);

		int m = X.size();
		int n = theta.size();

		for(int i = 0; i < n; i++)
		{
			ret.add(theta.get(i));
		}

		while(J > err)
		{
			ArrayList<Double> temp = new ArrayList<Double>();
			for(int j = 0; j < n; j++)
			{
				temp.add(ret.get(j));
				double theta_j = ret.get(j);
				for(int i = 0; i < m; i++)
				{
					theta_j -= (alpha/m)*(pderiv(j, ret, X, Y));
				}
				temp.set(j, theta_j);
			}
			for(int j = 0; j < n; j++)
			{
				ret.set(j, temp.get(j));
			}

			J = cost(ret, X, Y);
			cost_hist.add(J);

			if(cost_hist.get(cost_hist.size()-1)-cost_hist.get(cost_hist.size()-2) > err)
			{
				ret.set(0, Double.MAX_VALUE);
				break;
			}
		}

		return ret;
	}

	public static double cost(ArrayList<Double> theta, ArrayList<Double> X, ArrayList<Double> Y)
	{
		double J = 0;
		int m = X.size();
		for(int i = 0; i < m; i++)
		{
			J += Math.pow(polyEval(theta, X.get(i))-Y.get(i), 2);
		}
		J /= (2*((double) m));

		return J;
	}
	public static double polyEval(ArrayList<Double> theta, double x)
	{
		double eval = theta.get(0);
		for(int i = 1; i < theta.size(); i++)
		{
			eval += theta.get(i)*Math.pow(x, i);
		}
		return eval;
	}
	public static double pderiv(int j, ArrayList<Double> theta, ArrayList<Double> X, ArrayList<Double> Y)
	{
		double delta = 0;
		int m = X.size();
		if(j == 0)
		{
			for(int i = 1; i < m; i++)
			{
				delta += (polyEval(theta, X.get(i))-Y.get(i));
			}
		}
		else
		{
			for(int i = 1; i < m; i++)
			{
				delta += (polyEval(theta, X.get(i))-Y.get(i))*Math.pow(X.get(i), j);
			}
		}

		return delta/((double) m);
	}
}

File: func.py
import math
import csv
import argparse
import numpy as np
import matplotlib.pyplot as plt


def gradDescent(theta, X, Y, err, alpha):
    ret = []
    cost_hist = []
    J = float("inf")
    cost_hist.append(0.0)
    cost_hist.append(J)

    m = len(X)
    n = len(theta)

    for i in range(n):
        ret.append(theta[i])

    while(J > err):
        # Cauchy convergence : gradDescnet has converged to some other point but error is still larger
        # than what user desires. Thus, means we must use a higher order
        # polynomial.
        if(math.fabs(cost_hist[len(cost_hist) - 1] - cost_hist[len(cost_hist) - 2]) < err / 10000):
            ret[0] = 'bad'
            break

        temp = []
        for j in range(n):
            theta_j = ret[j]

            temp.append(theta_j)
            theta_j -= (alpha / m) * (pderiv(j, ret, X, Y))
            temp[j] = theta_j

        for j in range(n):
            ret[j] = temp[j]
        # print(ret)
        J = cost(ret, X, Y)
        # print(J)
        cost_hist.append(J)

    return ret


def cost(theta, X, Y):
    J = 0
    m = len(X)
    for i in range(m):
        J += ((polyEval(theta, X[i]) - Y[i])**2)
    J /= (2 * m)
    return J


def polyEval(theta, x):
    evaluate = theta[0]
    for i in range(len(theta)):
        evaluate += theta[i] * (x**i)
    return evaluate


def pderiv(j, theta, X, Y):
    # Partial derivative of cost function with respect to theta_j
    delta = 0
    m = len(X)

    if(j == 0):
        for i in range(1, m):
            delta += (polyEval(theta, X[i]) - Y[i])
    else:
        for i in range(1, m):
            delta += (polyEval(theta, X[i]) - Y[i]) * (X[i]**j)

    return delta / m

if __name__ == '__main__':
    parser = argparse.ArgumentParser(
        description='Produces a polynomial that best fits the given data set. Uses Gradient Descent algorithm.')
    parser.add_argument(
        'file', type=str, help='Must pass absolute path to a .csv file (formatted in two columns)')
    parser.add_argument(
        'err', type=float, help='Pass a floating point value that denotes the error threshold (Mean Square Error).')
    arg = parser.parse_args()

    f = arg.file
    err = arg.err

    X = []
    Y = []

    with open(f, 'rb') as csvfile:
        reader = csv.reader(csvfile, delimiter=',')
        for row in reader:
            X.append(float(row[0]))
            Y.append(float(row[1]))

    plt.scatter(X, Y)

    alpha = 0.0009  # Somewhat arbitrary learning rate
    theta = [0.0]

    coeff = gradDescent(theta, X, Y, err, alpha)
    while(coeff[0] == 'bad'):
        theta.append(0.0)
        coeff = gradDescent(theta, X, Y, err, alpha)

    plt.plot(X, [polyEval(coeff, x) for x in X])
    plt.show()

    print('Coefficients: ' + str(coeff))
    print('Cost: ' + str(cost(coeff, X, Y)))
    print('\n')
